From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Paolo Sabatino <paolo.sabatino@gmail.com>
Date: Thu, 16 Jun 2022 19:19:49 +0000
Subject: fix spreadtrum (sprd) bluetooth broken park link status

Spreadtrum/uwe5622 Bluetooth firmware reports Park link mode support but
rejects HCI commands to enable it. This patch adds HCI_QUIRK_BROKEN_PARK_LINK_STATUS.
---
 drivers/bluetooth/hci_ldisc.c | 6 ++++++
 include/net/bluetooth/hci.h   | 6 ++++++
 net/bluetooth/hci_sync.c      | 2 +-
 3 files changed, 13 insertions(+), 1 deletion(-)

diff --git a/drivers/bluetooth/hci_ldisc.c b/drivers/bluetooth/hci_ldisc.c
index 111111111111..222222222222 100644
--- a/drivers/bluetooth/hci_ldisc.c
+++ b/drivers/bluetooth/hci_ldisc.c
@@ -700,6 +700,12 @@ static int hci_uart_register_dev(struct hci_uart *hu)
 	hdev->setup = hci_uart_setup;
 	SET_HCIDEV_DEV(hdev, hu->tty->dev);
 
+	// Set the broken Park link status quirk, specific for spreadtrum (sprd)
+	// bluetooth devices
+	if (hdev->manufacturer == 0xffff && hu->tty->driver &&
+		strncmp(hu->tty->driver->name, "ttyBT", 5) == 0)
+		hci_set_quirk(hdev, HCI_QUIRK_BROKEN_PARK_LINK_STATUS);
+
 	if (test_bit(HCI_UART_RAW_DEVICE, &hu->hdev_flags))
 		hci_set_quirk(hdev, HCI_QUIRK_RAW_DEVICE);
 
diff --git a/include/net/bluetooth/hci.h b/include/net/bluetooth/hci.h
index 111111111111..222222222222 100644
--- a/include/net/bluetooth/hci.h
+++ b/include/net/bluetooth/hci.h
@@ -378,6 +378,12 @@ enum {
 	 */
 	HCI_QUIRK_BROKEN_READ_PAGE_SCAN_TYPE,
 
+	/*
+	 * Device declares that support Park link status, but it really
+	 * does not support it and fails to initialize
+	 */
+	HCI_QUIRK_BROKEN_PARK_LINK_STATUS,
+
 	__HCI_NUM_QUIRKS,
 };
 
diff --git a/net/bluetooth/hci_sync.c b/net/bluetooth/hci_sync.c
index 111111111111..222222222222 100644
--- a/net/bluetooth/hci_sync.c
+++ b/net/bluetooth/hci_sync.c
@@ -4261,7 +4261,7 @@ static int hci_setup_link_policy_sync(struct hci_dev *hdev)
 		link_policy |= HCI_LP_HOLD;
 	if (lmp_sniff_capable(hdev))
 		link_policy |= HCI_LP_SNIFF;
-	if (lmp_park_capable(hdev))
+	if (lmp_park_capable(hdev) && !hci_test_quirk(hdev, HCI_QUIRK_BROKEN_PARK_LINK_STATUS))
 		link_policy |= HCI_LP_PARK;
 
 	cp.policy = cpu_to_le16(link_policy);
-- 
Armbian

